src/Nethermind/Nethermind.AuRa/AuRaBlockFinalizationManager.cs

```diff
@@ -0,0 +1,269 @@
+
+            return finalizedBlocks;
+        }
+
+        /* Simple, unoptimized method implementation for reference: 
+        private IReadOnlyList<BlockHeader> GetFinalizedBlocks(BlockHeader block)
+        {
+            (ChainLevelInfo parentLevel, BlockInfo parentBlockInfo) GetBlockInfo(BlockHeader blockHeader)
+            {
+                var chainLevelInfo = _blockInfoRepository.LoadLevel(blockHeader.Number);
+                var blockInfo = chainLevelInfo.BlockInfos.First(i => i.BlockHash == blockHeader.Hash);
+                return (chainLevelInfo, blockInfo);
+            }
+
+            List<BlockHeader> finalizedBlocks = new List<BlockHeader>();
+            var validators = new HashSet<Address>();
+            var minSealersForFinalization = block.IsGenesis ? 1 : _auRaValidator.MinSealersForFinalization;
+            var originalBlockSealer = block.Beneficiary;
+            
+            using (var batch = _blockInfoRepository.StartBatch())
+            {
+                var (chainLevel, blockInfo) = GetBlockInfo(block);
+                
+                while (!blockInfo.IsFinalized)
+                {
+                    validators.Add(block.Beneficiary);
+                    if (validators.Count >= minSealersForFinalization)
+                    {
+                        blockInfo.IsFinalized = true;
+                        _blockInfoRepository.PersistLevel(block.Number, chainLevel, batch);
+                        finalizedBlocks.Add(block);
+                    }
+
+                    if (!block.IsGenesis)
+                    {
+                        block = _blockTree.FindHeader(block.ParentHash, BlockTreeLookupOptions.None);
+                        (chainLevel, blockInfo) = GetBlockInfo(block);
+                    }
+                }
+            }
+
+            finalizedBlocks.Reverse(); // we were adding from the last to earliest, going through parents
+            
+            return finalizedBlocks;
+        }
+        */
+
+        public event EventHandler<FinalizeEventArgs> BlocksFinalized;
+
+        public long LastFinalizedBlockLevel
+        {
+            get => _lastFinalizedBlockLevel;
+            private set
+            {
+                if (_lastFinalizedBlockLevel < value)
+                {
+                    _lastFinalizedBlockLevel = value;
+                    if (_logger.IsTrace) _logger.Trace($"Setting {nameof(LastFinalizedBlockLevel)} to {value}.");
+                }
+            }
+        }
+
+        public void Dispose()
+        {
+            _blockProcessor.BlockProcessed -= OnBlockProcessed;
+        }
+
+        [DebuggerDisplay("Count = {Count}")]
+        private class ValidationStampCollection
+        {
+            private readonly ISet<Address> _set = new HashSet<Address>();
+            private readonly SortedList<long, Address> _list = new SortedList<long, Address>(Comparer<long>.Create((x, y) => (int) (y - x)));
```

Comment:
is there no standard long comparer?

Reply:
also the comparison feels wrong -> shpuld be y - x > 0L ? 1 : y - x == 0L ? 0 : -1